_posts/2020-06-26-Making-IrishGen-SPARQL-Part-I.md

```diff
@@ -0,0 +1,549 @@
+| http://example.com/Rawl_B502/genelach_ceníuil_bóguine.trig#Báeth               | "Báeth"     |           |
+| http://example.com/Rawl_B502/do_primforslointib_Lagen_inso.trig#Báeth-604282d0 | "Báeth"     | "Báeth"   |
+| http://example.com/LL/de_genelach_dail_messi_corbb.trig#Baeth-a245b020         | "Baeth"     | "Báeth"   |
+| http://example.com/LL/de_genelach_dail_messi_corbb.trig#Baeth-a245b020         | "Báeth"     | "Báeth"   |
+| http://example.com/LL/genelach_h_n-enechglais.trig#Baeth-50d79733              | "Baeth"     |           |
+| http://example.com/LL/genelach_h_falgi.trig#Báeth                              | "Báeth"     |           |
+| http://example.com/Rawl_B502/genelach_benntraige.trig#Báeth                    | "Báeth"     | "Báeth"   |
+| http://example.com/Rawl_B502/clann_aingeda.trig#Báeth                          | "Báeth"     | "Báeth"   |
+| http://example.com/Rawl_B502/genelach_ceníuil_dalláin.trig#Báeth               | "Báeth"     |           |
+| http://example.com/Rawl_B502/genelach_úa_m-bairrche.trig#Báeth                 | "Báeth"     | "Báeth"   |
+| http://example.com/Rawl_B502/genelach_úa_m-bairrche.trig#Bóeth                 | "Baeth"     |           |
+
+In this case though, there is another thing to note.  The query will
+return all results which have a nominative name.  This could result in
+the tens of thousands.  It will then filter those by the regular
+expression and that will spawn tens or hundreds of thousands of
+computations or more.  While in this case it took only 1816
+millisecond (1.8 seconds), this is very inefficient.  Users will need
+to have some awareness of the complexity of their queries when running
+them and strive for the most specific query that will satisfy their
+requirements.  Some queries, however, will just take time and the user
+will need to keep in mind that their queries will not always be
+instantaneous.
+
+The next natural question is: what happens when both names could be
+present?  As the reader can see from the example above the `optional`
+keyword needs `{` and `}` which implies that more than one thing can
+be optional at a time.  The general case can be re-specified as: the
+subject can have either `irishRel:nomName` or `irishRel:genName` or
+both.  In this case SPARQL has the
+[`union`](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#alternatives)
+keyword which combines the results of graph patterns.  Thus, in the
+case where both `irishRel:nomName` or `irishRel:genName` can be
+optional, the query can be rewritten:
+
+```sparql
+prefix irishRel: <http://example.com/earlyIrishRelationship.ttl#> 
+
+select ?x ?y
+from <tag:stardog:api:context:all>
+where {
+    optional { 
+      {?x irishRel:nomName ?y }
+      union
+      { ?x irishRel:genName ?y } 
+    }
+    filter regex(?y, "^B[áa]eth$", "i")
+}
+```
+
+The above query will return 24 results.  This will not distinguish
```

Comment:
Again, I think we should have a discussion about this because I looked at `http://example.com/Rawl_B502/do_primforslointib_Lagen_inso.trig#Báeth-604282d0` and it is defined thus:

```turtle
<#Báeth-604282d0>
    a foaf:Person;
    irishRel:genName "Báeth";
    irishRel:nomName "Báeth";
    rel:childOf <#Ninnida>;
    owl:sameAs <#Báeth>.
```